src/main/scala/chisel3/util/LTL.scala

```diff
@@ -0,0 +1,392 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util.ltl
+
+import chisel3._
+import chisel3.util.circt._
+import chisel3.experimental.hierarchy.{Instance, Instantiate}
+
+/** An opaque sequence returned by an intrinsic.
+  *
+  * Due to the lack of opaque user-defined types in FIRRTL, the Linear Temporal
+  * Logic (LTL) intrinsics operate on `Bool` values instead of a more desirable
+  * `Sequence` type. To prevent abuse, the Chisel API here wraps these `Bool`s
+  * in an opaque type over which we can provide a safe interface that cannot be
+  * used to construct invalid IR.
+  */
+private case class OpaqueSequence(private[ltl] val inner: Bool) extends Sequence
+
+/** An opaque property returned by an intrinsic.
+  *
+  * See `OpaqueSequence` for details.
+  */
+private case class OpaqueProperty(private[ltl] val inner: Bool) extends Property
+
+/** A single item that may be used in the `Sequence(...)` convenience
+  * constructor for sequences. These atoms may either be `Sequence`s themselves,
+  * like `a` or `a and b`, or a `DelayAtom`, like `###`. Together they enable a
+  * shorthand notation for sequences like:
+  *
+  * `Sequence(a, ###, b, ###(2), c, ###(3, 9), d, ###(4, None))`.
+  */
+sealed trait SequenceAtom
+
+/** A delay atom for the `Sequence(...)` convenience constructor. * */
+private case class DelayAtom(val min: Int, val max: Option[Int]) extends SequenceAtom
+
+/** The delay atoms available to users. Can be interleaved with actual sequences
+  * in `Sequence(...)`. See `SequenceAtom` for details.
+  */
+object ### extends SequenceAtom {
+  def apply(delay: Int): SequenceAtom = DelayAtom(delay, Some(delay))
+  def apply(min:   Int, max: Int): SequenceAtom = DelayAtom(min, Some(max))
+  def apply(min:   Int, max: Option[Int]): SequenceAtom = DelayAtom(min, max)
+}
+
+/** A Linear Temporal Logic (LTL) sequence. */
+trait Sequence extends Property {
+
+  /** See `Sequence.delay`. */
+  def delay(delay: Int = 1): Sequence = Sequence.delay(this, delay)
+
+  /** See `Sequence.delayRange`. */
+  def delayRange(min: Int, max: Int): Sequence = Sequence.delayRange(this, min, max)
+
+  /** See `Sequence.delayAtLeast`. */
+  def delayAtLeast(delay: Int): Sequence = Sequence.delayAtLeast(this, delay)
+
```

Comment:
Are the values returned here actually valid `Sequences`? `a.delay(2)` doesn't seem like a valid Sequence, I think you need a "that" argument for these to make them valid or they should be returning a different type (maybe `SequenceFragment`?, that might be worse than just requiring another argument to complete the `Sequence`.

Reply:
Yeah, `a.delay(2)` is a valid sequence. It delays `a` by two cycles, corresponding to `##2 a` in SVA.

Reply:
SVA conflates concatenation and delay. The CIRCT LTL intrinsics separate them properly: concat is a variadic operation that just concatenates sequences, and delay is a unary operation on a single sequence. That's why the API here has them separate, but there is a convenience `###` operator (and the `Sequence(...)`) stuff that offers a combined concat-and-delay as you'd expect to see them in SVA, to make it easier to write sequences.

Reply:
Makes sense, thanks for the explanation!